File: marketplace/aac-skills/hooks/scripts/_plugin_hook_guard.js
// Shared dedup guard for governance hooks that ride in both the plugin payload
// and the live-tree ~/.claude/settings.json (issue 208 criterion 4).
//
// On a PC where the user-settings entries still point at the live-tree copy
// (~/.claude/hooks/<name>.js), the plugin's own hook entry would fire the SAME
// event a second time. This guard makes the plugin copy exit 0 silently when a
// live-tree copy is present -- so the hook fires once, from user-settings --
// while a container that has no live tree runs the plugin copy unimpeded.
//
// Once the paired live-tree/mirror edit ships (sync.ps1 -Mode push after
// removing the user-settings governance entries), the live-tree hooks script
// directory still exists but the settings entries are gone; that is fine, the
// guard only checks the file's presence to decide who owns the run. Set
// PLUGIN_HOOK_GUARD_DISABLE=1 in the env to force the plugin to run anyway
// (useful for local testing).
'use strict';

const fs = require('node:fs');
const os = require('node:os');
const path = require('node:path');

function callerFile() {
  // process.argv[1] is the script the harness invoked, which is what we care
  // about (the guard is a require()d dependency, __filename here is the guard).
  const argv1 = process.argv[1];
  if (argv1) {
    try { return fs.realpathSync(argv1); } catch (_) { return path.resolve(argv1); }
  }
  return null;
}

function pluginRootCanonical() {
  const r = process.env.CLAUDE_PLUGIN_ROOT;
  if (!r) return null;
  try { return fs.realpathSync(r); } catch (_) { return path.resolve(r); }
}

function isPluginCopy(caller, pluginRoot) {
  if (!caller || !pluginRoot) return false;
  const rel = path.relative(pluginRoot, caller);
  return rel && !rel.startsWith('..') && !path.isAbsolute(rel);
}

function liveTreeTwinExists(caller) {
  const configDir = process.env.CLAUDE_CONFIG_DIR || path.join(os.homedir(), '.claude');
  const live = path.join(configDir, 'hooks', path.basename(caller));
  try { return fs.statSync(live).isFile(); }
  catch (_) { return false; }
}

function skipIfLiveTreeWillFire() {
  if (process.env.PLUGIN_HOOK_GUARD_DISABLE === '1') return;
  const pluginRoot = pluginRootCanonical();
  if (!pluginRoot) return; // not a plugin invocation
  const caller = callerFile();
  if (!isPluginCopy(caller, pluginRoot)) return;
  if (!liveTreeTwinExists(caller)) return;
  // Silent exit -- Claude Code treats a 0-exit hook with no stdout as a no-op.
  process.exit(0);
}

module.exports = { skipIfLiveTreeWillFire };

File: marketplace/aac-skills/hooks/scripts/_plugin_hook_guard.py
"""Shared dedup guard for governance hooks that ride in both the plugin payload
and the live-tree ~/.claude/settings.json (issue 208 criterion 4).

On a PC where the user-settings entries still point at the live-tree copy
(~/.codex/hooks/ask_matt_gate.py), the plugin's own hook entry would fire the
SAME event a second time. This guard makes the plugin copy exit 0 silently when
user settings still dispatch this script -- so the hook fires once, from
user-settings -- while a container that has no such entry runs the plugin copy
unimpeded.

The test is "does settings.json name this script", NOT "does the live file
exist": the live-tree files stay after the paired edit lands (they are the
packager's source for this very payload and the restore test executes them),
so a presence check would keep skipping forever and the hook would fire zero
times. Only Claude Code's settings.json is consulted: ~/.codex/hooks.json
dispatches Codex sessions, which never load this plugin.

Set PLUGIN_HOOK_GUARD_DISABLE=1 in the env to force the plugin to run anyway.
"""
from __future__ import annotations

import json
import os
import sys
from pathlib import Path


def _caller_path(script_path: Path) -> Path:
    try:
        return script_path.resolve()
    except OSError:
        return script_path


def _plugin_root_canonical() -> Path | None:
    raw = os.environ.get("CLAUDE_PLUGIN_ROOT")
    if not raw:
        return None
    try:
        return Path(raw).resolve()
    except OSError:
        return Path(raw)


def _is_plugin_copy(caller: Path, plugin_root: Path) -> bool:
    try:
        caller.relative_to(plugin_root)
    except ValueError:
        return False
    return True


def _user_settings_dispatch(caller: Path) -> bool:
    """True when ~/.claude/settings.json carries a hook command naming this script's basename.

    Plugin-root commands never appear there, so a match is the live-tree entry and the live copy
    is the one that fires."""
    claude_home = Path(os.environ.get("CLAUDE_CONFIG_DIR") or Path.home() / ".claude")
    try:
        settings = json.loads((claude_home / "settings.json").read_text(encoding="utf-8"))
    except (OSError, ValueError):
        return False
    events = settings.get("hooks") if isinstance(settings, dict) else None
    if not isinstance(events, dict):
        return False
    name = caller.name
    for groups in events.values():
        if not isinstance(groups, list):
            continue
        for group in groups:
            hooks = group.get("hooks") if isinstance(group, dict) else None
            for hook in hooks or []:
                if not isinstance(hook, dict):
                    continue
                for key in ("command", "commandWindows"):
                    cmd = hook.get(key)
                    if isinstance(cmd, str) and name in cmd:
                        return True
    return False


def skip_if_live_tree_will_fire(script_path: Path) -> None:
    if os.environ.get("PLUGIN_HOOK_GUARD_DISABLE") == "1":
        return
    plugin_root = _plugin_root_canonical()
    if plugin_root is None:
        return  # not a plugin invocation
    caller = _caller_path(script_path)
    if not _is_plugin_copy(caller, plugin_root):
        return
    if not _user_settings_dispatch(caller):
        return
    sys.exit(0)
